### PR TITLE
GitHub desktop

### PR DESCRIPTION
**Pull request by theofficialgman**
armhf and arm64
Do not merge until this exists in pi-apps-coders files first
https://github.com/Pi-Apps-Coders/files/pull/29

yes I made this from github desktop on my switch

### CODE REVIEW

**Comment by ryanfortner** [comment]
@theofficialgman could you send the instructions to get it built please?

**Comment by theofficialgman** [comment]
> @theofficialgman could you send the instructions to build please?

as I described over here https://github.com/shiftkey/desktop/issues/251#issuecomment-1118176428 https://github.com/shiftkey/desktop/issues/251#issuecomment-1118884407 

its a bit complicated, I still need to make some .patch files of the dugite-native and desktop repos so it can be replicated by others.

**Comment by theofficialgman** [comment]
@ryanfortner I would suggest using my deb. this has the most compatibility the same as upstream github (ubuntu bionic and newer)

regardless here are the instructions:

Patch for dugite-native(https://github.com/desktop/dugite-native):
https://github.com/theofficialgman/dugite-native/commit/b73556385a470d1c26826862f8bce52f7ecf3d12.patch
from the main repo folder, apply the above patch first:
```
npm install
script/build.sh
script/package.sh
```
this will give you a file named close to `dugite-native-v2.35.3-eda48e6-ubuntu.tar.gz`

patch for shiftkey github desktop (https://github.com/shiftkey/desktop):
https://github.com/theofficialgman/desktop/commit/96e313319b0b88eca86045719851665c98bf3a8d.patch

all from the main folder of the repo. apple the patch above first:
```
yarn
yarn lint
```
then extract the dugite-native tar.gz into `app/node_modules/dugite` so that the you now have a `git` folder in that dugite folder
continue from the main repo folder:
```
yarn build:prod
yarn run package
```
you should now have the packages in the `dist` folder

keep in mind none of these changes are upstreamable, and are just meant for making the package locally. upstream builds on github actions where native arm64 builds are not easily (or quickly which is the problem) possible

**Comment by Crilum** [comment]
Could this be built on armhf?

**Comment by theofficialgman** [comment]
> Could this be built on armhf?

not sure yet.. I'm still making my bionic armhf debootstrap. if all the npm natives have armhf binaries then yes.. but if they don't its going to be a deep rabbit hold of making them

**Comment by theofficialgman** [comment]
got my bionic armhf chroot made and my dependencies and updated npm and jarn repos added. I've built armhf dugite-native already (it its basically gitweb, git, curl, and a few other things all smashed together). github desktop is a much longer build... we will see how that goes

I've updated my instructions a bit, I forgot about a couple of steps before.

**Comment by theofficialgman** [comment]
as I expected github-desktop is proving harder.. but not for the reason I expected. npm is installing the windows dependencies even though I'm running on linux (even though they were never used). this went fine on arm64 since there is an arm64 windows release. there is no npm arm32 windows support in most things so its failing. so now I'm having to purge the windows crap to build armhf linux.

**Comment by theofficialgman** [comment]
I have built ARMhf Github Desktop 🎉 

Incoming deb, appimage, and rpm @Crilum @ryanfortner @Botspot needs testing, same location
https://github.com/theofficialgman/testing/releases/tag/github-desktop

its too late in the day for me to document what additional changes were necessary. I'd post patch files tomorrow or another day

**Comment by Crilum** [comment]
@theofficialgman On armhf I've signed in, and have opened a repo, it looks and feels really smooth.

I probably won't be able to do extensive testing till tomorrow, but so far, looks good!

**Comment by Crilum** [comment]
Ok, one error so far:

In the `Repository` menu, there is an option to `Open in GNOME Terminal`, and since by default RPiOS doesn't have GNOME Terminal installed, GH Desktop outputs an error saying: `Cannot read properties of undefined (reading 'path')`. It works after I install `gnome-terminal` though.

**Comment by Crilum** [comment]
After a few minutes of just looking through the menus, GH Desktop randomly seg-faulted...

**Comment by Crilum** [comment]
> After a few minutes of just looking through the menus, GH Desktop randomly seg-faulted...

I think this was a random occurrence, and it doesn't need to be worried about.

**Comment by Crilum** [comment]
The armhf AppImage works fine as well...

File: apps/Github Desktop/install-32 [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+install_packages https://github.com/Pi-Apps-Coders/files/raw/main/GitHubDesktop-linux-armhf-3.0.0.deb || exit 1
```
Comment (Crilum) [inline apps/Github Desktop/install-32:3]: To fix GNOME Terminal error:
```suggestion
install_packages https://github.com/Pi-Apps-Coders/files/raw/main/GitHubDesktop-linux-armhf-3.0.0.deb gnome-terminal || exit 1
```

File: apps/Github Desktop/install-64 [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+install_packages https://github.com/Pi-Apps-Coders/files/raw/main/GitHubDesktop-linux-arm64-3.0.0.deb || exit 1
```
Comment (Crilum) [inline apps/Github Desktop/install-64:3]: To fix GNOME Terminal error:
```suggestion
install_packages https://github.com/Pi-Apps-Coders/files/raw/main/GitHubDesktop-linux-arm64-3.0.0.deb gnome-terminal || exit 1
```

File: apps/Github Desktop/description [hunk 3]
```diff
@@ -0,0 +1,5 @@
+Github Desktop is a git GUI client.
+Focus on what matters instead of fighting with Git. Whether you're new to Git or a seasoned user, GitHub Desktop simplifies your development workflow.
+
+Run from Menu: Menu -> Development -> Github Desktop
```
Comment (Crilum) [inline apps/Github Desktop/description:4]: Desktop appears in `Accessories` for me on RPiOS.

File: apps/Github Desktop/description [hunk 4]
```diff
@@ -0,0 +1,5 @@
+Github Desktop is a git GUI client.
```
Comment (Crilum) [inline apps/Github Desktop/description:1]:
```suggestion
Github Desktop is a GUI git client.
```

Comment (Itai-Nelken) [reply apps/Github Desktop/description:1]: It's more a github gui client than a git one.

**Comment by theofficialgman** [comment]
> Ok, one error so far:
> 
> In the `Repository` menu, there is an option to `Open in GNOME Terminal`, and since by default RPiOS doesn't have GNOME Terminal installed, GH Desktop outputs an error saying: `Cannot read properties of undefined (reading 'path')`. It works after I install `gnome-terminal` though.
> 
> When this gets added to Pi-Apps, `gnome-terminal` should be added as a dependency.

Yeah this is just due to how github desk

**Comment by Jai-JAP** [comment]
> > Ok, one error so far:
> > 
> > In the `Repository` menu, there is an option to `Open in GNOME Terminal`, and since by default RPiOS doesn't have GNOME Terminal installed, GH Desktop outputs an error saying: `Cannot read properties of undefined (reading 'path')`. It works after I install `gnome-terminal` though.
> > 
> > When this gets added to Pi-Apps, `gnome-terminal` should be added as a dependency.
> 
> Yeah this is just due to how github desktop is coded. The list of supported terminals is small. I know xterm, gnome-terminal, and konsole is in that list. What terminal does piOS have installed by default?

PiOS comes preinstalled with lxterminal

**Comment by theofficialgman** [comment]
Ok here is the list of supported terminals.
https://github.com/shiftkey/desktop/blob/linux/app/src/lib/shells/linux.ts
The list is actually quite large but lxterminal is not there and I guess you didn't have any of the others installed

**Comment by Jai-JAP** [comment]
Well them u need to add lxterminal for lxde but lxde is now dead. LxQt is the new successor for lxde

**Comment by theofficialgman** [comment]
> Well them u need to add lxterminal for lxde but lxde is now dead. LxQt is the new successor for lxde

LXDE is still actively developed. LXQt is a separate project. LXDE is gtk based. LXQt is QT based. both very different gui libraries

**Comment by theofficialgman** [comment]
@Jai-JAP which one looks better themed on piOS lxde, gnome-terminal or xterm?

**Comment by Jai-JAP** [comment]
For editor its geany on PiOS. and mousepad on twister os tou need to add them both.

And lxterminal looks good on PiOS when themed but i haven't tried other terminals

**Comment by Jai-JAP** [comment]
For editor its geany on PiOS. and mousepad on twister os tou need to add them both.

And lxterminal looks good on PiOS when themed but i haven't tried other terminals
Also xterm is the worst in terms of looks even themed. 😕

**Comment by theofficialgman** [comment]
ok then I'll just add geany and lxterminal upstream to github desktop and hold back this PR till merged

**Comment by theofficialgman** [comment]
on second though, lets just add gnome-terminal and gedit for now to the depends so this can be merged.
I'll still PR to upstream and we can rebuild once that is merged

**Comment by Jai-JAP** [comment]
> on second though, lets just add gnome-terminal and gedit for now to the depends so this can be merged ^.
> I'll still PR to upstream and we can rebuild once that is merged

Wait until someone confirms shell name on PiOS. I'm not sure if its LXDE or PIXEL or something else.
I may do it tommorrow

**Comment by theofficialgman** [comment]
> > on second though, lets just add gnome-terminal and gedit for now to the depends so this can be merged ^.
> > I'll still PR to upstream and we can rebuild once that is merged
> 
> Wait until someone confirms shell name on PiOS. I'm not sure if its LXDE or PIXEL or something else. I may do it tommorrow

lxterminal is the name of the shell on LXDE, piOS uses LXDE so I don't expect them to change the name

**Comment by theofficialgman** [comment]
@Itai-Nelken updated the debs so that lxterminal and geany are supported.
see PR: https://github.com/Pi-Apps-Coders/files/pull/30

**Comment by Jai-JAP** [comment]
> PR opened for geany and lxterminal:
> https://github.com/shiftkey/desktop/pull/711
> 
> I have built arm64 binaries but I can't upload them right now. will do later today

@theofficialgman  also add mousepad to supported editors as twister os uses it by default.

**Comment by theofficialgman** [comment]
> @theofficialgman also add mousepad to supported editors as twister os uses it by default.

I'll add it to the PR for github desktop. I won't rebuild since that takes to much time for little benefit, but whenever github desktop 3.1 releases I'll rebuild then

**Comment by Itai-Nelken** [comment]
> @Itai-Nelken @Botspot updated the debs so that lxterminal and geany are supported for both armhf and arm64
> see PR: https://github.com/Pi-Apps-Coders/files/pull/30
> 
> once that is merged, this can be merged as well

Merged.

**Comment by theofficialgman** [comment]
@Botspot can you add the bitly links and then I'll merge?

**Pull request merged**